### PR TITLE
Fix `TabBar` tab indicator stretch effect

### LINKED ISSUE #149662: [Material3] TabBar indicator stretch effect behaving weirdly with long tabs
### Steps to reproduce

Create a TabBar with some text tabs in different lengths, even with some that are truncated. TabBar can be scrollable or not (but issue is more prominent with scrollable).

### Expected results

The TabBar indicator should animate smoothly even with tabs with longer text, without expanding in opposite direction of slide animation.

### Actual results

Indicator first expands to the left (if navigating to the right) while extending width of indicator. This expanding to the left does not seem to happen on shorter tabs and causes the animation to look glitchy. Also if swiping quickly between screens/tabs this becomes very jarring.

The new `_applyStretchEffect` method used in `tabs.dart` seems to be making the indicator behave weirdly when tabs with longer text is used. I tried setting `final double stretchSize = rect.width;` to half or a quarter, which seems to help with the expanding width animation a bit, but it's not perfect.

The best solution (for the project I'm working on at least) would be to have `_applyStretchEffect` as an optional parameter, as disabling it works best for the scrolling use case (according to me). But this might not go in line with the material spec?

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MainApp());
}

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: DefaultTabController(
        length: 8,
        child: SafeArea(
          child: Scaffold(
            appBar: TabBar(
              isScrollable: true,
              tabAlignment: TabAlignment.center,
              tabs: [
                Tab(text: 'Home'),
                Tab(text: 'Search'),
                Tab(text: 'Add'),
                Tab(text: 'Favorite'),
                Tab(text: 'The longest text...'),
                Tab(text: 'Short'),
                Tab(text: 'Longer text...'),
                Tab(text: 'Profile'),
              ],
            ),
            body: TabBarView(
              children: <Widget>[
                Center(child: Text("Page")),
                Center(child: Text("Page")),
                Center(child: Text("Page")),
                Center(child: Text("Page")),
                Center(child: Text("Page")),
                Center(child: Text("Page")),
                Center(child: Text("Page")),
                Center(child: Text("Page")),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

Example of jarring animation:

https://github.com/flutter/flutter/assets/35582133/779758cb-4f8d-44d8-b08f-36907ccaf348

https://github.com/flutter/flutter/assets/35582133/7059989a-6f33-4c7b-979c-9ffeffec2656

With `final double stretchSize = rect.width / 2;`

https://github.com/flutter/flutter/assets/35582133/77a246fa-4e17-43c7-af99-92d86b0d0ca1

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.22.1, on macOS 14.5 23F79 darwin-arm64, locale en-SE)
    • Flutter version 3.22.1 on channel stable at /Users/<user>/fvm/versions/3.22.1
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision a14f74ff3a (13 days ago), 2024-05-22 11:08:21 -0500
    • Engine revision 55eae6864b
    • Dart version 3.4.1
    • DevTools version 2.34.3

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/<user>/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/<user>/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15C65
    • CocoaPods version 1.15.2

[✗] Chrome - develop for the web (Cannot find Chrome executable at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[✓] Android Studio (version 2023.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)

[✓] VS Code (version 1.89.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.90.0

[✓] VS Code (version 1.83.0-insider)
    • VS Code at /Applications/Visual Studio Code - Insiders.app/Contents
    • Flutter extension version 3.72.0

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

</details>

### PR DESCRIPTION
**Pull request by TahaTesser**
fixes [[Material3] TabBar indicator stretch effect behaving weirdly with long tabs](https://github.com/flutter/flutter/issues/149662)

This PR fixes the tab indicator stretch effect, it currently stretches from both sides.  After the fix, tab indicator stretches depending on next size of the indicator and direction of the scroll.

The fix is based on Android Components implementation of the elastic/stretch effect.

https://github.com/material-components/material-components-android/blob/20f92dfb513916d0df6c38bfd630dfc41aff484e/lib/java/com/google/android/material/tabs/ElasticTabIndicatorInterpolator.java#L46-L78

### Code sample

<details>
<summary>expand to view the code sample</summary> 

```dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: ScrollConfiguration(
        behavior: ScrollConfiguration.of(context)
            .copyWith(dragDevices: <PointerDeviceKind>{
          PointerDeviceKind.touch,
          PointerDeviceKind.mouse,
        }),
        child: DefaultTabController(
          length: 8,
          child: Scaffold(
            appBar: AppBar(
              bottom: const TabBar(
                isScrollable: true,
                tabAlignment: TabAlignment.start,
                tabs: <Widget>[
                  Tab(text: 'Home'),
                  Tab(text: 'Search'),
                  Tab(text: 'Add'),
                  Tab(text: 'Favorite'),
                  Tab(text: 'The longest text...'),
                  Tab(text: 'Short'),
                  Tab(text: 'Longer text...'),
                  Tab(text: 'Profile'),
                ],
              ),
            ),
            body: const TabBarView(
              children: <Widget>[
                Center(child: Text('Page')),
                Center(child: Text('Page')),
                Center(child: Text('Page')),
                Center(child: Text('Page')),
                Center(child: Text('Page')),
                Center(child: Text('Page')),
                Center(child: Text('Page')),
                Center(child: Text('Page')),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
```

</details>

### Before

https://github.com/flutter/flutter/assets/48603081/618d0ba9-40a5-458f-9fdc-5330505a6711

### After

https://github.com/flutter/flutter/assets/48603081/b7fa851e-e7a6-4b66-b77d-f88c7f4381da

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/lib/src/material/tabs.dart [hunk 1]
```diff
@@ -575,8 +575,18 @@ class _IndicatorPainter extends CustomPainter {
     _painter!.paint(canvas, _currentRect!.topLeft, configuration);
   }
 
+  // Ease out sine (decelerating).
+  double decelerateIntepolation(double fraction) {
```
Comment (Piinks) [inline packages/flutter/lib/src/material/tabs.dart:579]:
```suggestion
  double decelerateInterpolation(double fraction) {
```

File: packages/flutter/lib/src/material/tabs.dart [hunk 2]
```diff
@@ -575,8 +575,18 @@ class _IndicatorPainter extends CustomPainter {
...
+    return math.sin((fraction * math.pi) / 2.0);
+  }
+
+  // Ease in sine (accelerating).
+  double accelerateIntepolation(double fraction) {
```
Comment (Piinks) [inline packages/flutter/lib/src/material/tabs.dart:584]:
```suggestion
  double accelerateInterpolation(double fraction) {
```

**Review by Piinks — COMMENTED** [review]
In the after video it looks like there is a slight glitch in the indicator just after is finishes animating to the long tab. After it arrives, it looks like it really quickly shortens the indicator length on the right side. Or is that maybe triggered by the next swipe gesture starting?

**Comment by TahaTesser** [comment]
@Piinks 
Great observation, the issue might be that i used to trackpad to test in the videos. I'll put some real device comparison and see if it really glitches.

**Comment by TahaTesser** [comment]
> @Piinks Great observation, the issue might be that i used to trackpad to test in the videos. I'll put some real device comparison and see if it really glitches.

I tested this closely on desktop and mobile. This happens when you tap the before the tab indicator indications is completed and swipe tab animation is interrupted. This also happens without this PR too, it's not a bug from this PR.

**Comment by TahaTesser** [comment]
In the next PR. adding a way customize this stretch/elastic tab indicator animation so user can choose to use linear tab indicator animation as part of another issue. https://github.com/flutter/flutter/issues/150508


https://github.com/flutter/flutter/assets/48603081/de147c28-d9ab-4c45-89a1-927cec443d5f

File: packages/flutter/lib/src/material/tabs.dart [hunk 3]
```diff
@@ -586,57 +596,37 @@ class _IndicatorPainter extends CustomPainter {
...
+    final double rightFraction;
 
-    final double inflationPerSide = stretchSize * _stretchAnimation.transform(tabChangeProgress) / 2;
-    final Rect stretchedRect = _inflateRectHorizontally(rect, inflationPerSide);
+    if (isMovingRight) {
+      leftFraction = accelerateInterpolation(tabChangeProgress);
+      rightFraction = accelerateInterpolation(tabChangeProgress);
+    } else {
+      leftFraction = decelerateInterpolation(tabChangeProgress);
+      rightFraction = decelerateInterpolation(tabChangeProgress);
```
Comment (Piinks) [inline packages/flutter/lib/src/material/tabs.dart:615]: If these are just the same values for right and left in both cases, why not use one variable?

Comment (Piinks) [reply packages/flutter/lib/src/material/tabs.dart:615]: I also thought of using a switch statement + pattern combo for assignment, but I figure that implementation preference is not really important.  🙂

**Review by Piinks — APPROVED** [review]
This LGTM with a nit on the duped variables - you can choose to ignore if you favor it for readability!

**Pull request merged**